### PR TITLE
Archive workspace chats optimistically

### PR DESCRIPTION
**Pull request by aldo-expensify**
### Details

Missed to optimistically archive the workspace chats and rooms when refactored the Workspace Delete.

Without archiving optimistically, if the user is slow or with a slow connection, they can delete the workspace and then keep queueing messages on the workspace's rooms and chats until we receive the pusher event archiving them. Now, the chats and rooms get optimistically archived and, on failure, we restore their original state/status.

### Fixed Issues

$ https://github.com/Expensify/Expensify/issues/215184#issuecomment-1258452400

### Tests

1. Create expensifail account and validate it
2. Create workspace
3. Send a message in every chat and room of the workspace: `#announce`, `#admins` and the policy expense chat
4. Open the workspace settings
5. In Chrome dev tools, network tab, disable the internet connection
6. Delete the workspace:
    image
7. You should see the workspace strikethrough as deletion is pending and the chat and rooms archived:
    image
8. Enable the internet connection
9. Check that the chats remain archived and the workspace gets deleted
10. Only if you have access to the Web-API code:
    1. Add a throw line like this `throw new ExpError('1F036B68-0C7E-4AC2-9E5A-A0A0E835C447', 'An unexpected error occured while trying to delete the workspace.');` [here](https://github.com/Expensify/Web-Expensify/blob/306789746b756f95d5b98ece623398c4eebb9fb6/lib/PolicyAPI.php#L4092) (inside the `try` block and right before the call to `self::delete`)
    2. Create a workspace and try to delete it
    3. Check that on failure, the chats and the rooms status is restored

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps
1. Create expensifail account and validate it
2. Create workspace
3. Send a message in every chat and room of the workspace: `#announce`, `#admins` and the policy expense chat
4. Open the workspace settings
5. In Chrome dev tools, network tab, disable the internet connection
6. Delete the workspace:
    image
7. You should see the workspace strikethrough as deletion is pending and the chat and rooms archived:
    image
8. Enable the internet connection
9. Check that the chats remain archived and the workspace gets deleted

- [x] Verify that no errors appear in the JS console

#### Web

Try to delete workspace offline, see optimistic archiving of the chats. Get back online, API throws a hardcoded error. See un-archiving of chats by `failureData`

https://user-images.githubusercontent.com/87341702/193925641-bd77ea1c-8383-4dfb-8f1d-f30048ac757a.mov

#### Mobile Web - Safari

https://user-images.githubusercontent.com/87341702/193928964-028ac331-f84a-4141-9e6d-156b8d22f2b1.mp4

https://user-images.githubusercontent.com/87341702/193929509-db244a08-160f-498f-abb4-b44880dbdd43.mp4

#### Desktop

Try to delete workspace offline, see optimistic archiving of the chats. Get back online, API throws a hardcoded error. See un-archiving of chats by `failureData`

image
image

#### iOS

https://user-images.githubusercontent.com/87341702/193945186-ecbc28f6-2644-42ac-9505-07bda6b63a63.mp4

This is an workspace's employee getting the push notification to archive the policy expense chat and remove the `#announce` room after the owner (from another device) deletes the workspace

https://user-images.githubusercontent.com/87341702/193946728-1b6548b5-8260-4f19-9116-567e75c5997d.mp4

#### Android
Try to delete workspace, the request is delayed in the API so we can see the optimistic changes. The API throws an error and the optimistic changes are reverted (chats get unarchived and the workspace is there again)

https://user-images.githubusercontent.com/87341702/194617337-acd4c3f8-c81c-44ff-b1ae-28e408bc9756.mp4

This is an workspace's employee getting the push notification to archive the policy expense chat and remove the `#announce` room after the owner (from another device) deletes the workspace

https://user-images.githubusercontent.com/87341702/193952388-e70c8d2c-0051-400f-a8dc-ecc4787a3443.mp4

### CODE REVIEW

**Comment by aldo-expensify** [comment]
cc @roryabraham since you brought up the incomplete optimistic behaviour. I have some doubt if the optimistically archived chats are ok like that or if we should do something with "pending actions" where they are strikethrough and not archived while the request has not resolved yet.

**Review by roryabraham — DISMISSED** [review]
This LGTM. Definitely an improvement over what we had.

Unless I'm mistaken, I don't think we need to worry about archiving "pending chats" because in the scenario we're talking about, those chats would be sent and become non-pending before the workspace is actually deleted:

1. I am offline
1. I send a chat in the `#admins` room like "Hey, I'm deleting this workspace because we don't need it anymore". This is my "pending chat"
1. I optimistically delete the workspace, and we optimistically archive the `#admins` room.
1. Then I come back online, and my serialized requests happen in order (thanks to the `SequentialQueue`):
    1. My chats are sent in the `#admins` room
    1. Then the workspace is deleted and the `#admins` room is archived
    
So it wouldn't really make sense to strikethrough my pending chats in the `#admins` room in step `3`, because they'll end up being sent after all.

**Comment by aldo-expensify** [comment]
@roryabraham I had not really thought about the case you described, and what you said makes sense to me on that case. Sorry for the confusion, but I was really talking about a strikethrough of the room itself (not pending chats inside it) because it is "in the process of being archived but it haven't resolved yet".

**Comment by roryabraham** [comment]
Oh, no I think changing it to the archived state like you do here is fine. After the reports switch to being archived though, we should make sure that we can't send messages in them (the compose bar should be replaced with a message that says the room is archived)

**Comment by aldo-expensify** [comment]
Update: added check to make sure the code doesn't crash if there is a `[reportID-onyx-key] => undefined` in the Onyx reports collection... which I think can happen when reports are removed (`Onyx.set(reportID-onyx-key, null)`).

**Comment by aldo-expensify** [comment]
> After the reports switch to being archived though, we should make sure that we can't send messages in them (the compose bar should be replaced with a message that says the room is archived)

Yep, I checked that, if you open the optimistically archived report it looks like any archived report and you cannot send new messages

**Review by stitesExpensify — DISMISSED** [review]

**Comment by roryabraham** [comment]
Found a bit of an edge case:

image

**Comment by roryabraham** [comment]
Created a separate issue: https://github.com/Expensify/Expensify/issues/231206

File: src/pages/workspace/WorkspaceInitialPage.js [hunk 1]
```diff
@@ -35,6 +35,21 @@ const propTypes = {
 
     /** The employee list of this policy (coming from Onyx) */
     policyMemberList: PropTypes.objectOf(policyMemberPropType),
+
+    /** All reports shared with the user (coming from Onyx) */
+    reports: PropTypes.shape({
+        /** The report name */
+        reportID: PropTypes.number,
```
Comment (roryabraham) [inline src/pages/workspace/WorkspaceInitialPage.js:42]: This just changed to `string` everywhere: https://github.com/Expensify/App/pull/11362

```suggestion
        reportID: PropTypes.string,
```

Comment (roryabraham) [reply src/pages/workspace/WorkspaceInitialPage.js:42]: Also, this propType doesn't seem right. Should it be:

```js
reports: PropTypes.objectOf(reportPropTypes),
```

Where `reportPropTypes` comes from [this new file on main](https://github.com/Expensify/App/blob/main/src/pages/reportPropTypes.js)?

**Review by roryabraham — CHANGES_REQUESTED** [review]
Just a minor propTypes comment, otherwise LGTM

**Comment by trjExpensify** [comment]
> Found a bit of an edge case:
> 
> image

Commented in that issue, but I think this PR is introducing that regression as this is all working as intended on staging. Check [the comment for the screenies and run down](https://github.com/Expensify/Expensify/issues/231206#issuecomment-1263411327).

File: src/pages/workspace/WorkspaceInitialPage.js (hunk 1 above)
Comment (aldo-expensify) [reply src/pages/workspace/WorkspaceInitialPage.js]: Updated

**Comment by aldo-expensify** [comment]
> Found a bit of an edge case:
> 
> image

I think this is expected if the employee somehow missed the pusher event archiving the chats.

**Comment by JmillsExpensify** [comment]
Dang, what's our plan for getting to the bottom of this Pusher notification missses?

**Comment by trjExpensify** [comment]
Yeah, if someone else can't reproduce that issue on staging, we should close it. It was working fine for me!

**Review by roryabraham — APPROVED** [review]

**Comment by JmillsExpensify** [comment]
> @JmillsExpensify In my opinion, that doesn't affect this pull request review, because we should only be looking at the optimistic onyx update behaviour and what happens on failure (revert the optimistic changes). This is not affecting the pusher events sent to other employees or the owner of the workspace, it only affects what the owner sees on delete when it has a slow/non-existent connection.

Nice, that makes sense. Thanks for the breakdown. 

> If the pusher events in dev are buggy and the code is fine, we should close this other GH issue: https://github.com/Expensify/Expensify/issues/231206

Do ya'll want me to test on staging? I can, though I think it's probably more productive to ensure this isn't a dev issue first.

**Comment by aldo-expensify** [comment]
Merging since it was approved by @roryabraham, and @stitesExpensify is OOO. I don't know why the `PR Author Checklist / checklist check` is failing, I'm pretty sure I checked everything.

As a reminder, the comments left on the PR and this other GH issue: https://github.com/Expensify/Expensify/issues/231206 shouldn't be related to the changes here.

**Comment by aldo-expensify** [comment]
Removed emergency, the Author and reviewer checklist checks were failing. I tried checking all Author related checks, but still was failing.. maybe there is a change in the template?

**Comment by aldo-expensify** [comment]
QA: ✅ 


https://user-images.githubusercontent.com/87341702/195464458-7a080215-29cd-48eb-9385-2d7546b1927b.mov

**Pull request merged**